### PR TITLE
Add Index v2 documentation

### PR DESCRIPTION
**Pull request by geoffturk**

### CODE REVIEW

File: index_api-v2.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
+# Murmurations Index API
+
+> :link: BASE URL
+>
+> _Test Environment_  
+> https://test-index.murmurations.network2/{endpoint}
```
Comment (Photosynthesis) [inline index_api-v2.md:6]: `https://test-index.murmurations.network2/{endpoint}` should be `https://test-index.murmurations.network/v2/{endpoint}`

File: index_api-v2.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+Nodes store their profiles on their website or through a dedicated profile host and then request them to be referenced in the index. The index validates that the profile meets the requirements of its associated schema(s). The index does not actually store the entire profile; it is only required to store the URL of the profile's location on the web (`profile_url`) and the name(s) of the schema(s) that profile is/are based upon and must be validated against (`linked_schemas`).
+
+Using these two pieces of information, aggregators can then search the index for nodes with profiles that match specific schemas they want to use to create maps and directories, and the index returns a list of URLs (`profile_url`s) for the matched nodes' profiles. Aggregators then download from each node it's profile data for use in their apps.
+
+The index also stores additional data to enable aggregators to locate nodes based on other information in addition to their associated schemas. This additional information includes, for example:
```
Comment (Photosynthesis) [inline index_api-v2.md:19]: My suggestion as an edit for the above three paragraphs (suggesting this because it avoids the split between some index fields and others, which is a bit obsolete at this point, and adds the primary URL).

Nodes store their profiles on their website or through a dedicated profile host and then request them to be referenced in the index. The index validates that the profile meets the requirements of its associated schema(s). The index does not store the entire profile, only certain fields that can then be used by aggregators to find profiles that match various search criteria. 

Information stored by the index includes, for example:
 - _Linked schemas_ - the names of the schemas that the profile has been built and validated against
 - _Profile URL_ - the URL where the profile is stored
 - _Primary Node URL - the website address of the node

File: index_api-v2.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+<!-- TODO - add this back in if/when a solution is determined
+- _Entity type_ - Wikipedia/Wikidata URL that best describes the organization type - enables searching for specific types of entities (e.g., food co-ops: [https://en.wikipedia.org/wiki/Food_cooperative](https://en.wikipedia.org/wiki/Food_cooperative))
+-->
+
+> :construction: INDEX SYNCING
+>
+> A `profile_hash` is for future planning, and will also be stored for each node profile so that multiple indices can synchronize their lists of nodes. It will be a hash of the profile (the entire JSON object containing all node profile data) stored at the `profile_url` that was submitted by the node. Nodes can compare `node_id`s (the unique hash of a `profile_url`) and then `profile_hash`es between themselves. They will also store a Unix timestamp (`last_updated`) of when they last validated the profile at the `profile_url` and created the `profile_hash` for each `node_id`. The index with the oldest timestamp should download the profile from the `profile_url` and regenerate the `profile_hash`. This is just a rough idea of index syncing; it will need to be thought through in more detail, including thinking about edge cases.
```
Comment (Photosynthesis) [inline index_api-v2.md:30]: Since we're getting into the reality of multiple indices, and the requirement to synchronize, merge, and deduplicate profiles, it might be a good time to revisit this and think about it in more detail.

File: index_api-v2.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+## Node Endpoints
+
+### [`POST /validate`](https://app.swaggerhub.com/apis-docs/MurmurationsNetwork/IndexAPI/2.0.0#/Node%20Endpoints/post_validate)
+
+Node operators can test the validity of a profile to one or more schemas before posting the profile to their website and submitting the `profile_url` to the index.
```
Comment (Photosynthesis) [inline index_api-v2.md:36]: I was a bit sidetracked by "to one or more schemas" here -- made me think there should be an input parameter for the schema(s).

File: index_api-v2.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+#### Input
+
+- A valid JSON schema instance (the profile).
```
Comment (Photosynthesis) [inline index_api-v2.md:40]: I'd suggest we change this to something like: "The profile JSON to validate, including an array of one or more `linked_schemas` to validate against.

File: index_api-v2.md [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+#### Output
+
+- 200 OK response (if profile validates to the schema(s))
+- 400 Bad Request response (if profile does not validate) along with one or more failure reasons
```
Comment (Photosynthesis) [inline index_api-v2.md:45]: It may be too late to change this in the current round of updates, but I don't think we should use 400 responses in this case. This should return a 200 response, because the purpose of this endpoint is to check a profile that may or may not be valid. A request to this endpoint that tells the node that the profile is invalid is still a successful request (it achieved the purpose of the endpoint). A 400 error, according to MDN:

"indicates that the server cannot or will not process the request due to something that is perceived to be a client error (for example, malformed request syntax, invalid request message framing, or deceptive request routing)." 

In our case, for this endpoint an invalid profile is not an invalid request in any of the senses above. 

On the practical side, using 4xx responses can cause problems on the client, because systems that manage the HTTP(S) communications rightly treat these responses as indications of an error at the request level that should never occur, and treat them as exceptions that need to be handled using a different code flow than normal results. As in the case of the previous situation of 404s for empty search results, awkward coding is required to work around this.

File: index_api-v2.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+> :construction: HASH ALGORITHM
+>
+> The current implementation of the index uses the SHA256 hashing algorithm which produces a 64-character output. The purpose of hashing the `profile_url` is to make it easy to reference as a path parameter when requesting information about the node from the index (e.g., `GET /nodes/{node_id}` as described below).
+
+#### Sequence Diagram - Add Node Profile to Index
+
+![Add Node Profile to Index](index_api-v1-diagram_1.png)
+
+### [`POST /nodes-sync`](https://app.swaggerhub.com/apis-docs/MurmurationsNetwork/IndexAPI/2.0.0#/Node%20Endpoints/post_nodes_sync)
```
Comment (Photosynthesis) [inline index_api-v2.md:83]: Awesome to have this!

File: index_api-v2.md [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+#### Input
+
+- the `node_id` of the node's profile that is currently in the index
+
+#### Output
+
+Detailed response examples can be seen in the [Open API specification](https://app.swaggerhub.com/apis-docs/MurmurationsNetwork/IndexAPI/2.0.0#/Node%20Endpoints/get_nodes__node_id_) under the _Responses_ section for `200 OK` response code.
+
+##### Success
```
Comment (Photosynthesis) [inline index_api-v2.md:109]: Looks to me like maybe these success and error outputs should be under the POST /nodes endpoint, rather than GET /nodes ?

File: index_api-v2.md [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+- Could not post validated profile to the index (internal service issue)
+
+### [`DELETE /nodes/{node_id}`](https://app.swaggerhub.com/apis-docs/MurmurationsNetwork/IndexAPI/2.0.0#/Node%20Endpoints/delete_nodes__node_id_)
+
+Node operators will use the `DELETE /nodes/{node_id}` endpoint to remove their profile from the index when they no longer want it listed.
+
+#### Input
+
+- the `node_id` of the profile that is currently in the index
+  - The profile must no longer be available at the `profile_url` (i.e., should return a `404 Not Found` error when accessing the URL) or it will not be removed from the index
```
Comment (Photosynthesis) [inline index_api-v2.md:132]: Because systems don't necessarily do a good job of returning 404s for missing resources (ie, when they're using URL rewrites to show a custom 404 page, or other redirection), I think `delete` requests should also take effect when the profile URL doesn't return a 404, but returns something else that's not valid JSON. (I know we discussed this at some point, and I think it might be already implemented. If not, maybe we add it to the backlog for future updates).

File: index_api-v2.md [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+## Aggregator Endpoints
+
+### [`GET /nodes`](https://app.swaggerhub.com/apis-docs/MurmurationsNetwork/IndexAPI/2.0.0#/Aggregator%20Endpoints/get_nodes)
+
+<!-- TODO - add this back if/when it happens
+> :lock: The `GET /nodes` endpoint will eventually require an API key in order to prevent unauthorized harvesting of data from the index.
+ -->
+
+Aggregators can search for nodes based on the schemas nodes use, when the nodes were last validated by the index (i.e., for recent changes to node profiles), and by geolocating nodes within a certain range (e.g., _10km_ or _6mi_) from a specific location or finding them based on the town/city, country, etc. Searching for deleted nodes is also possible so that aggregators can keep their records updated as nodes remove themselves from the index. Finally, nodes can associate tags to their profiles to enable searching by specific keywords.
```
Comment (Photosynthesis) [inline index_api-v2.md:153]: Suggestion (last_validated > updated, and adding primary_url)

Aggregators can search for nodes based on the schemas nodes use, when the nodes were last updated, and by geolocating nodes within a certain range (e.g., _10km_ or _6mi_) from a specific location or finding them based on the town/city, country, etc. Searching for deleted nodes is also possible so that aggregators can keep their records updated as nodes remove themselves from the index. Finally, nodes can associate tags to their profiles to enable searching by specific keywords, and nodes can be queried based on their `primary_url` value.

File: index_api-v2.yaml [hunk 11]
```diff
@@ -0,0 +1,722 @@
...
+            schema:
+              $ref: "#/components/schemas/ValidateNode"
+            example:
+              linked_schemas:
+                - "demo_schema-v1"
+              name: "Some Org"
+      responses:
+        200:
+          description: OK
+        400:
```
Comment (Photosynthesis) [inline index_api-v2.yaml:43]: See comment above re using 400 for this...

File: index_api-v2.yaml [hunk 12]
```diff
@@ -0,0 +1,722 @@
...
+          $ref: "#/components/responses/ServerError"
+    get:
+      tags:
+        - Aggregator Endpoints
+      summary: Search for nodes
+      description: |
+        Aggregators can search for nodes based on any of the following parameters:
+
+        - A specific schema (`schema`)
+        - When the node was last validated by the index (`last_updated`)
```
Comment (Photosynthesis) [inline index_api-v2.yaml:128]: Shouldn't this be: "When the node was last updated (`last_updated`)" ? I assume this won't be reset when, for example, the `validate` endpoint is requested.

File: index_api-v2.yaml [hunk 13]
```diff
@@ -0,0 +1,722 @@
...
+        - A distance in kilometers (_"25km"_) or miles (_"15mi"_)) from a specific geolocation (`lat`, `lon`, `range`)
+        - By city/town/village/etc, state/province/county/etc. and/or country. (`locality`, `region`, `country`)
+        - By node profile status (`posted` or `deleted`)
+        - By tags that describe the node (using an AND/OR filter with fuzzy or exact matching)
+        - By the nodes website address (`primary_url`)
+
+        Query results can be paginated using the `page` and `page_size` parameters.
+
+        Aggregators will eventually be required to authenticate themselves with a JWT to access this endpoint.
```
Comment (Photosynthesis) [inline index_api-v2.yaml:137]: Are we sure that JWT is the technology we want to use for this? I wonder if it's overkill for aggregator authentication (unless there's something we need to do that won't work with a simple API key approach).

File: index_api-v2.yaml [hunk 14]
```diff
@@ -0,0 +1,722 @@
...
+                      profile_hash: "c24d14c2c75f55d334a7e0ccf4d35a063a2582a7abb91e16d326f6613b9602bf"
+                posted:
+                  value:
+                    data:
+                      node_id: "a55964aeaae9625dc2b8dbdb1c4ce0ed1e658483f44cf2be1a6479fe5e144d38"
+                      profile_url: "https://somenode.org/optional-subdirectory/node-profile.json"
+                      status: "posted"
+                      last_updated: 1601979232403
+                      profile_hash: "c24d14c2c75f55d334a7e0ccf4d35a063a2582a7abb91e16d326f6613b9602bf"
+                deleted:
```
Comment (Photosynthesis) [inline index_api-v2.yaml:250]: This probably won't happen, no?

File: index_api-v2.yaml [hunk 15]
```diff
@@ -0,0 +1,722 @@
...
+                      failure_reasons:
+                        - "Could not read from schema: example_schema-v1"
+                post_failed:
+                  value:
+                    data:
+                      node_id: "a55964aeaae9625dc2b8dbdb1c4ce0ed1e658483f44cf2be1a6479fe5e144d38"
+                      profile_url: "https://somenode.org/optional-subdirectory/node-profile.json"
+                      status: "post_failed"
+                      last_updated: 1601979232403
+                NodeNotFoundInIndex:
```
Comment (Photosynthesis) [inline index_api-v2.yaml:376]: Let's talk about how we use response codes for errors at different layers in the API. Mentioning here because I think this would be a legit place to use a 404 (since a specific resource was requested and not found). But, I'm happy that it's not using a 404, because I'm in favor of using HTTP error responses only for situations where the request to the index was fundamentally broken (ie, non-existing endpoint, missing required parameters for the endpoint, rate limits, authentication, etc.) I think we could think this through a bit, and see what sort of consistent pattern we could settle on.

File: index_api-v2.yaml [hunk 16]
```diff
@@ -0,0 +1,722 @@
...
+        500:
+          $ref: "#/components/responses/ServerError"
+    delete:
+      tags:
+        - Node Endpoints
+      summary: Delete a node from the index
+      description: |
+        A node can delete its profile from the index at any time simply by removing the profile from its `profile_url` on its website and then sending this DELETE request to the index.
+
+        The index will first confirm the profile is no longer available at the `profile_url` (node's website should return a `404 - Not Found` error) and then remove the profile from its records.
```
Comment (Photosynthesis) [inline index_api-v2.yaml:390]: See comment above re sites that don't necessarily return a 404 when they should because of redirection.

File: index_api-v2.yaml [hunk 17]
```diff
@@ -0,0 +1,722 @@
...
+    schema:
+      name: schema
+      in: query
+      description: The name of the schema
+      schema:
+        type: string
+    last_updated:
+      name: last_updated
+      in: query
+      description: Unix timestamp in seconds when node was last validated by index
```
Comment (Photosynthesis) [inline index_api-v2.yaml:589]: See comment above re validated vs updated. My assumption is that this will be the (self reported) update time of the node, rather than the last time the index validated the node (for example via a /validate request)

File: index_api-v2.md (hunk 1 above)
Comment (geoffturk) [reply index_api-v2.md:6]: Good catch. Thanks!

File: index_api-v2.md (hunk 2 above)
Comment (geoffturk) [reply index_api-v2.md:19]: I've incorporated your suggestion. Please see my next commit and let me know if its OK.

File: index_api-v2.md (hunk 3 above)
Comment (geoffturk) [reply index_api-v2.md:30]: OK. I created an issue for this:

https://github.com/MurmurationsNetwork/MurmurationsProtocol/issues/40

File: index_api-v2.md (hunk 4 above)
Comment (geoffturk) [reply index_api-v2.md:36]: As you noted below, there is the `linked_schemas` parameter. I also changed "to one or more schemas" to "against one or more schemas" as it is better English.

File: index_api-v2.md (hunk 5 above)
Comment (geoffturk) [reply index_api-v2.md:40]: Much better. I have made the change.

File: index_api-v2.md (hunk 6 above)
Comment (geoffturk) [reply index_api-v2.md:45]: OK. I updated the docs to reflect this and created an issue:
https://github.com/MurmurationsNetwork/MurmurationsServices/issues/265

We'll make the change before we release v2 into production.

File: index_api-v2.md (hunk 7 above)
Comment (geoffturk) [reply index_api-v2.md:83]: 🚀

File: index_api-v2.md (hunk 8 above)
Comment (geoffturk) [reply index_api-v2.md:109]: No, this is correct. The `POST /nodes` endpoint only returns a 200 OK to confirm that the `profile_url` has been received along with a `node_id`. It is then put into a queue and processed, and the status of the processing can be found using the `GET /nodes/{node_id}` endpoint.

More explanation on this point below in response to your other related questions...

File: index_api-v2.yaml [hunk 18]
```diff
@@ -0,0 +1,722 @@
...
+                  data:
+                    $ref: "#/components/schemas/PostNodeSyncResponse"
+              examples:
+                received:
+                  value:
+                    data:
+                      node_id: "a55964aeaae9625dc2b8dbdb1c4ce0ed1e658483f44cf2be1a6479fe5e144d38"
+                      profile_url: "https://somenode.org/optional-subdirectory/node-profile.json"
+                      status: "received"
+                validated:
```
[the remark this answers is not in the document]
Comment (geoffturk) [reply index_api-v2.yaml:234]: Yes, if there are thousands of new `profile_urls` sent to the Index all at once, there will be a queue and it may take a few seconds/minutes to process all of them, depending on the volume.

File: index_api-v2.yaml (hunk 14 above)
Comment (geoffturk) [reply index_api-v2.yaml:250]: What won't happen? Please elaborate.

File: index_api-v2.yaml (hunk 15 above)
Comment (geoffturk) [reply index_api-v2.yaml:376]: OK. I flagged it as a discussion point when we speak next.

File: index_api-v2.md (hunk 9 above)
Comment (geoffturk) [reply index_api-v2.md:132]: This is an important point, and the system does not handle broken JSON as "deleted". I've created an issue for it:
https://github.com/MurmurationsNetwork/MurmurationsServices/issues/266

We'll fix this prior to v2 going into production.

File: index_api-v2.md (hunk 10 above)
Comment (geoffturk) [reply index_api-v2.md:153]: Done.

File: index_api-v2.yaml (hunk 11 above)
Comment (geoffturk) [reply index_api-v2.yaml:43]: Noted, and fixed. See my next commit.

File: index_api-v2.yaml (hunk 12 above)
Comment (geoffturk) [reply index_api-v2.yaml:128]: Yes, that is correct. The `validate` endpoint knows nothing about the `profile_url` because the API user is sending the actual JSON profile as a payload to that endpoint, not the `profile_url`.

File: index_api-v2.yaml (hunk 13 above)
Comment (geoffturk) [reply index_api-v2.yaml:137]: No, it was just a suggestion. I updated the wording to reflect both and that it is to be determined in the future.

File: index_api-v2.yaml (hunk 16 above)
Comment (geoffturk) [reply index_api-v2.yaml:390]: Noted. I preemptively used the word "should" here (not "must") for a reason. ;-)

File: index_api-v2.yaml (hunk 17 above)
Comment (geoffturk) [reply index_api-v2.yaml:589]: See my previous comment about the `/validate` endpoint having no relation to this timestamp.

This timestamp is only updated each time the `POST /nodes` or `POST /nodes-sync` endpoints are sent the `profile_url` of the node in question. If nothing has changed, the timestamp is currently still updated to the moment the Index performed the validation work again. 

I suppose we could make a change that checks if the content of the `profile_url` is exactly the same, and if so, the timestamp remains as it was. Let me know if you think this is worth implementing.

File: index_api-v2.md (hunk 8 above)
Comment (Photosynthesis) [reply index_api-v2.md:116]: Right. Makes sense. I was a bit thrown off thinking of the GET /nodes endpoint as a more typical endpoint for retrieving an item, and the error or success info here is referring to the results of the POST, not the GET. But, since I suppose this endpoint will only really be used to check the status of a node (rather than to actually get it for something), this makes sense.

File: index_api-v2.yaml (hunk 18 above)
[the remark this answers is not in the document]
Comment (Photosynthesis) [reply index_api-v2.yaml:241]: Okay. My understanding was that this endpoint would only return once the posting of the profile was complete (so the client would be able to receive a conclusive answer about whether the node was posted successfully). What controls whether this is the case, for example if there are a large number of profiles posted at once?

File: index_api-v2.yaml (hunk 14 above)
Comment (Photosynthesis) [reply index_api-v2.yaml:257]: Sorry -- that was rather cryptic! Is there a scenario in which the POST /nodes-sync endpoint would return a status `deleted`?

File: index_api-v2.yaml (hunk 17 above)
Comment (Photosynthesis) [reply index_api-v2.yaml:603]: I think my comment was a bit unclear in which line it refers to. I was just thinking that the text "Unix timestamp in seconds when node was last validated by index" maybe should be changed to "Unix timestamp in seconds when node was last updated", because the node might get validated by the index without updating this value.

File: index_api-v2.yaml (hunk 18 above)
[the remark this answers is not in the document]
Comment (geoffturk) [reply index_api-v2.yaml:241]: Ah, OK, I see what you mean. Yes, you are right. I need to take out the the `received` and `validated` states from this list.

I think `post_failed` might still need to stay but I will test this and make sure.

File: index_api-v2.yaml (hunk 14 above)
Comment (geoffturk) [reply index_api-v2.yaml:257]: Good point. It probably should not. But there might be an edge case if the content of the profile_url changes from valid to invalid JSON and then the profile_url is sent again to the `/node-sync` endpoint. I need to test this out while making the change to treat invalid JSON responses as 404s.

File: index_api-v2.yaml (hunk 17 above)
Comment (geoffturk) [reply index_api-v2.yaml]: I fixed the wording, but my question above still stands. Does it make sense to leave this `last_updated` value unchanged if the `profile_url` is resent to `POST /nodes` (or `POST /nodes-sync`) but it's profile content is unchanged? Currently it the timestamp will be updated regardless.

Comment (Photosynthesis) [reply index_api-v2.yaml]: Good question. I think for now it makes sense to leave it as is, rather than comparing the content, but we should think about different scenarios and how this is used.

File: index_api-v2.yaml (hunk 18 above)
[the remark this answers is not in the document]
Comment (geoffturk) [reply index_api-v2.yaml]: There are edge cases where it is still possible that `received`, `validated` and `post_failed` can be returned by the `POST /nodes-sync` endpoint. They're rare, and the system would be in a half broken state (some services down), but it is possible so they should remain documented.

Only the `deleted` status is not possible to return so I have removed that in the docs.

File: index_api-v2.yaml (hunk 14 above)
Comment (geoffturk) [reply index_api-v2.yaml]: In the end, this won't happen so the `delete` example has been removed.
